### DISCUSSION TITLE
Spatialreference or equivilent for Disv-grid (mf6)

### DISCUSSION THREAD

**Discussion by d-strom** [pypest/pyemu#467] [Q&A] [answered]
Hi,

I've been struggling a bit with the pstfrom module since it needs information about the grid dimensions by use of the function SpatialReference.

As this functionality revolves around a structured x, y-grid approach I cant come up with a way to set it up for my unstructured models.

I have looked into some older threads on the topic where there are suggestions of an approach; 
https://github.com/pypest/pyemu/issues/246
https://github.com/pypest/pyemu/blob/e1ce35a34fa7c766f72f2a28b946be30c468bc13/autotest/pst_from_tests.py#L2840 

However, it feels a bit overcomplicated, especially for me whos not really fluent in python (or programming for that matter).

Since the thread is from 2021, I was wondering if anyone has come up with a more elegant and ready to go solution since then?

**Comment by jtwhite79** [comment, marked as answer]
We are actively working on pypestutils ([https://github.com/pypest/pypestutils](https://github.com/pypest/pypestutils)) and it has native support for unstructured grids in realm of both parameters and model post processing (like mod2obs, etc).  Its not quite ready-to-go, but Ive got a few examples and some helpers setup on my fork (will be merged into the main repo soon) that work on a quadtree variant of the Freyeberg model and it seems to be working.

**Reply by d-strom** [reply]
Great, thanks for the info. Looks like a very solid option for what I want to achieve. I'll look into it.

//David

**Reply by d-strom** [reply]
I ran some tests on my data with pypestutils but I'm running into a problem...

![image](https://github.com/pypest/pyemu/assets/74509984/8436eb58-2178-4d95-89d9-7a5e0a1dbb68)

As the image shows, I'm facing issues when flopy tries to load the disv-file. The file in my model is generated through modelmuse rather than flopy. But looking at them (attatched) I think they follow the same structure. The only difference being how layers are considerd as separate datasets in the pypestutils-example.

I'll attatch both disv-files to show how they are structured 

freyberg:

`# file generated by flopy version 3.3.6 on 12/30/2022 at 13:26:49.
begin options
  length_units  meters
  angrot -55
end options

begin dimensions
  nlay  3
  ncpl  5108
  nvert  6196
end dimensions

begin griddata
  top
    open/close  'freyberg6.disv_top.txt'  factor  1.0
  botm  layered
    open/close  'freyberg6.disv_botm_layer1.txt'  factor  1.0
    open/close  'freyberg6.disv_botm_layer2.txt'  factor  1.0
    open/close  'freyberg6.disv_botm_layer3.txt'  factor  1.0
  idomain  layered
    open/close  'freyberg6.disv_idomain_layer1.txt'  factor  1
    open/close  'freyberg6.disv_idomain_layer2.txt'  factor  1
    open/close  'freyberg6.disv_idomain_layer3.txt'  factor  1
end griddata

begin vertices
    open/close  'freyberg6.disv_vertices.txt'
end vertices

begin cell2d
    open/close  'freyberg6.disv_cell2d.txt'

end cell2d
`

my example-model

Attatched as a csv to fit the message.
[mod0000.csv](https://github.com/pypest/pyemu/files/13165506/mod0000.csv)

**Reply by jtwhite79** [reply]
@DStrom1987 Im not sure what is happening with that disv stuff - Im not an unstructured grid person.  But pypestuils can get grid info it needs from the MODFLOW-6 binary grid file (grb) (assuming you set the `xorigin`, `yorigin`, and `angrot` options in the disv package)...so you dont need flopy to use pypestutils...feel free to open issues and discussions on pypestutils as well!

**Comment by d-strom** [comment]
Thanks, I'm kind of a python newbie, so I really still need examples to
work with.

Atleast it makes things much easier. Which function would i use in
pypestutils for this operation and how would I call it?

Sincerely
David

Den ons 25 okt. 2023 21:57J Dub ***@***.***> skrev:

**Reply by jtwhite79** [reply]
The develop branch of pypestuils has some [examples](https://github.com/pypest/pypestutils/tree/develop/examples).  Admittedly, we need to make a new release so that the latest version of pypestutils is available to work with these examples (unless you are ok building the library yourself? - [build instructions](https://github.com/pypest/pypestutils/blob/main/BUILD.md)).  Otherwise, we are planning a release in the next couple of days...stay tuned!

**Comment by d-strom** [comment]
Of course!
Perhaps that's why the functions I'm calling are not recognized. I think
I'll wait for the release since I dont have so much time on my hands in the
coming days anyway.

Sincerely

David

Den ons 25 okt. 2023 kl 22:53 skrev J Dub ***@***.***>:
